Continue running kube-state-metrics when config file doesnt exist at startup

rashmichandrashekar opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:
1. If this is your first time, read our contributor guidelines https://git.k8s.io/community/contributors/guide/pull-requests.md#the-pull-request-submit-process and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
3. Follow the instructions for writing a release note: https://git.k8s.io/community/contributors/guide/pull-requests.md#write-release-notes-if-needed
4. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

**What this PR does / why we need it**:
This PR enables k-s-m should to run when config file doesnt exist.
Currently, [k-s-m expects the file provided in the config parameter to exist at startup](https://github.com/kubernetes/kube-state-metrics/blob/56d3b561e6954e0055ea0d6f2d7034f6d898b6c6/internal/wrapper.go#L50C1-L56C47) and if not, errors out. We are planning to take dependency on the config file by mounting it as an optional configmap which users can use to optionally override k-s-m configuration.

**How does this change affect the cardinality of KSM**: *(increases, decreases or does not change cardinality)*
Does not change cardinality

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #2700

mrueg commented on internal/wrapper.go:
```diff
@@ -43,17 +43,23 @@ func RunKubeStateMetricsWrapper(opts *options.Options) {
-				klog.ErrorS(err, "Options configuration file not found", "file", file)
+				klog.InfoS("Options configuration file not found at startup", "file", file)
+			} else if _, err = os.Stat(filepath.Clean(file)); os.IsNotExist(err) {
+				// Adding this check in addition to the above since viper.ConfigFileNotFoundError is not working as expected due to this issue -
```
```suggestion
				// TODO: Remove this check once viper.ConfigFileNotFoundError is working as expected, see this issue -
```

mrueg commented on pkg/app/server.go:
```diff
@@ -125,29 +125,33 @@ func RunKubeStateMetrics(ctx context.Context, opts *options.Options) error {
+
+			if err != nil {
+				// DO NOT end the process.
+				// We want to allow the user to still be able to fix the misconfigured config (redeploy or edit the configmaps) and reload KSM automatically once that's done.
```
I'm not sure if this is the behaviour we want from ksm. I'd rather prefer it failing if there's a bad config supplied.

rashmichandrashekar replied:
Thanks @mrueg. That behavior was not introduced in this PR, but I am happy to make the change and return error if needed. Although, I am unsure of the original intention of why this was implemented this way.

CatherineF-dev replied:
>  I'm not sure if this is the behaviour we want from ksm. I'd rather prefer it failing if there's a bad config supplied.

Agree

rashmichandrashekar replied:
Agree that it might be a breaking change.

rashmichandrashekar commented:
Gentle ping to all reviewers. 
@dgrisonnet @logicalhan @mrueg @rexagod - Please take a look and provide your feedback, thanks!

rashmichandrashekar commented:
@rexagod - Could you pls point me to the where the metrics need to be added? Thanks!

rashmichandrashekar commented:
Ping on this PR. Maintainers @richabanker @rexagod @mrueg @CatherineF-dev  - pls et me know the next steps and changes needed to proceed.

richabanker commented on internal/wrapper.go:
```diff
@@ -65,13 +71,15 @@ func RunKubeStateMetricsWrapper(opts *options.Options) {
+			}
 
-		yaml.Unmarshal(configFile, opts)
+			yaml.Unmarshal(configFile, opts)
```
oh this is previous behavior too, nvm.

richabanker commented:
I guess my last thoughts on this are that with this change, the application will continue to run even if the config file is missing, which is a substantial shift in user experience (users who rely on the application's exit code to validate their deployments may find their checks no longer work as expected) and should be called out in the release notes if we do decide to go forward with this.

rexagod commented:
> If it helps with backward compatibility I could also make it configurable with a cli parameter and the new behavior of continuing to run, could be set with this so that there is no breaking change.

I'm sorry, but why do we need this exactly? Introducing a breaking feature and putting that behind a flag, defaulting it to off, can set an unwanted precedent for the project. I don't mean to block you here, but we cannot introduce something like this unless there's a general consensus that the addition goes beyond downstream applications and helps the larger community.

rexagod commented:
Given how we have live-reloading for both KSM and CRS configurations, I can see the two complementing each other. However, wouldn't you need to do the same for CRS configuration as well (to deliver similar UX)? Would it make sense to have flags for both of these with the flag descriptions also highlighting the use-case for their inception?

rashmichandrashekar commented:
> Given how we have live-reloading for both KSM and CRS configurations, I can see the two complementing each other. However, wouldn't you need to do the same for CRS configuration as well (to deliver similar UX)? Would it make sense to have flags for both of these with the flag descriptions also highlighting the use-case for their inception?

Yes. Currently we dont support CR configuration. So, I didnt explore that. But I can make the change consistent across the two. I can add comments about the use cases for the config. So, just double checking before I implement - suggestion is to add a config flag that enables this behavior to provide for backward compatibility and make it consistent across KSM and CR configuration with comments.

rexagod commented:
Sounds good.

To reiterate a bit, it'd be nice to have a condensed version of the scenario you went over in your previous comment, in that flag(s) definition ([here](https://github.com/kubernetes/kube-state-metrics/blob/main/pkg/options/options.go#L143-L179)) as well.

rashmichandrashekar commented:
> Sounds good.
> 
> To reiterate a bit, it'd be nice to have a condensed version of the scenario you went over in your previous comment, in that flag(s) definition ([here](https://github.com/kubernetes/kube-state-metrics/blob/main/pkg/options/options.go#L143-L179)) as well, so other users know exactly when to use it (and thus deduce any similar use cases).

@rexagod - Addressed to add support for both and added description, pls check. Thanks!

rashmichandrashekar commented:
/ok-to-test

rashmichandrashekar commented:
Thanks @mrueg for triggering the test. I am running into issues with the cli-arguments. Is there a way to auto generate this file from the options file?

rashmichandrashekar commented:
/ok-to-test

rashmichandrashekar commented:
@rexagod - Addressed all the comments, pls take a look. We could squash it with merge into one, pls let me know if you want me to do it explicitly for this branch.

rashmichandrashekar commented:
/ok-to-test

richabanker commented:
/ok-to-test

rashmichandrashekar commented:
> Could you please squash the commits here as well? Thanks!

Done.

rashmichandrashekar commented on pkg/app/server.go:
```diff
@@ -539,11 +546,17 @@ func resolveCustomResourceConfig(opts *options.Options) (customresourcestate.Con
-		if err != nil {
-			return nil, fmt.Errorf("unable to open Custom Resource State Metrics file: %v", err)
+		if opts.ContinueWithoutCustomResourceConfigFile {
+			if _, err := os.Stat(filepath.Clean(file)); os.IsNotExist(err) {
```
@rexagod - I have updated this to log the error. pls let me know if you wanted me to do it another way. Since we want to continue to run, we can log the message and the error.

rashmichandrashekar commented on pkg/app/server.go:
```diff
@@ -596,11 +604,17 @@ func resolveCustomResourceConfig(opts *options.Options) (customresourcestate.Con
-		f, err := os.Open(filepath.Clean(file))
-		if err != nil {
-			return nil, fmt.Errorf("unable to open Custom Resource State Metrics file: %v", err)
+		if opts.ContinueWithoutCustomResourceConfigFile {
+			if _, err := os.Stat(filepath.Clean(file)); os.IsNotExist(err) {
+				klog.ErrorS(err, "failed to open Custom Resource State Metrics file, ignoring", "file", file)
+			}
+		} else {
```
This is how it shows up with the current implementation (but happy to update this to the suggestion, if needed) - 
<img width="1861" height="409" alt="image" src="https://github.com/user-attachments/assets/69438da6-0c07-4bad-946d-7a588b7e9063" />

rexagod commented on pkg/app/server.go:
```diff
@@ -178,15 +183,18 @@ func RunKubeStateMetrics(ctx context.Context, opts *options.Options) error {
-		crcFile, err := os.ReadFile(filepath.Clean(opts.CustomResourceConfigFile))
-		if err != nil {
-			return fmt.Errorf("failed to read custom resource config file: %v", err)
+		if _, err := os.Stat(filepath.Clean(opts.CustomResourceConfigFile)); err != nil {
+			klog.ErrorS(err, "config file does not exist, ignoring", "file", opts.CustomResourceConfigFile)
```
```suggestion
		if _, err := os.Stat(filepath.Clean(opts.CustomResourceConfigFile)); err != nil {
			klog.ErrorS(err, "encountered error while processing the file", "file", opts.CustomResourceConfigFile)
```

rexagod commented on pkg/app/server.go:
```diff
@@ -125,31 +125,36 @@ func RunKubeStateMetrics(ctx context.Context, opts *options.Options) error {
-			configSuccess.WithLabelValues("config", filepath.Clean(got)).Set(0)
-			<-ctx.Done()
+		if _, err := os.Stat(filepath.Clean(got)); err != nil {
+			klog.ErrorS(err, "config file does not exist, ignoring", "file", got)
```
```suggestion
			klog.ErrorS(err, "encountered error while processing the file", "file", got)
```

rexagod reviewed (commented):
One final nit regarding the error messages to reflect your last changes in the `if-else` conditionals. I believe we should be good to go after this.

rashmichandrashekar commented:
/ok-to-test

rashmichandrashekar commented:
> > One final nit regarding the error messages to reflect your last changes in the `if-else` conditionals. I believe we should be good to go after this.
> 
> @rexagod - Sure, made the changes, please check. Thanks!

@rexagod - Gentle ping, please let me know if this is good to merge or if there is my other feedback. Thanks!

rashmichandrashekar commented on internal/wrapper.go:
```diff
@@ -43,17 +43,25 @@ func RunKubeStateMetricsWrapper(opts *options.Options) {
-				klog.ErrorS(err, "Options configuration file not found", "file", file)
+		var cfgViperReadInConfigErr error
+		if cfgViperReadInConfigErr = cfgViper.ReadInConfig(); cfgViperReadInConfigErr != nil {
+			if errors.Is(cfgViperReadInConfigErr, viper.ConfigFileNotFoundError{}) {
```
@mrueg - This doesn't work as expected because of this bug - https://github.com/spf13/viper/issues/1783
The else if detects file doesn't exist reliably.

mrueg replied:
What I wanted to point at is the following comment:

```
errors.Is is intended for comparing against sentinel error variables (like os.ErrNotExist), not error types or struct values.
viper.ConfigFileNotFoundError{} is a struct, not a sentinel variable
```

rashmichandrashekar replied:
@mrueg - Thanks for pointing that out. The first if condition that you are pointing to always returns false currently because of the bug mentioned in the comment. We can fix that to also address your comment, when the original issue is fixed and remove the else if completely (please note this was something that existed before). The being said, I did test this out locally to make sure the flow works as expected.

rashmichandrashekar replied:
@mrueg - please let me know if this looks good or if you have additional concerns, thank you! It’s waiting on your review to merge

rexagod commented:
/lgtm
/hold

Holding since @mrueg's reviews are ongoing, feel free to uphold once they are addressed.

rashmichandrashekar commented on pkg/app/server.go:
```diff
@@ -596,11 +604,17 @@ func resolveCustomResourceConfig(opts *options.Options) (customresourcestate.Con
-			return nil, fmt.Errorf("unable to open Custom Resource State Metrics file: %v", err)
+		if opts.ContinueWithoutCustomResourceConfigFile {
+			if _, err := os.Stat(filepath.Clean(file)); err != nil {
+				klog.ErrorS(err, "failed to open Custom Resource State Metrics file, ignoring", "file", file)
```
Thanks @mrueg. Updated this, pls take a look

rashmichandrashekar replied:
@mrueg - Gentle ping, pls let me know if this looks good. Thanks!

rashmichandrashekar commented:
/ok-to-test

ringerc commented on internal/wrapper.go:
```diff
@@ -65,25 +73,33 @@ func RunKubeStateMetricsWrapper(opts *options.Options) {
+			}
 
-		yaml.Unmarshal(configFile, opts)
+			yaml.Unmarshal(configFile, opts)
```
Somehow this has broken in this patch. See https://github.com/kubernetes/kube-state-metrics/issues/2803#issuecomment-3464511667 . In 2.17.0 a config file containing

```
custom_resource_config_file: custom-resource-config.yaml
```

results in an `opts.CustomResourceConfigFile` that's the empty string.

[In 2.16.0 it instead gets the correct value.](https://github.com/kubernetes/kube-state-metrics/issues/2803#issuecomment-3464541147)
